backend/data/contractors.js

```diff
@@ -59,91 +33,143 @@ async function createContractor(
+  return contractor.inProgress;
 };
 
-const startNextTaskInQueue = async (contractorId) => {
+const startNextTaskInQueue = async (contractorId, projectId) => {
+  validation.checkNumOfArgs(arguments, 2);
+  validation.checkIsProper(contractorId, "string", "contractorId");
   validation.checkId(contractorId);
+  validation.checkIsProper(projectId, "string", "projectId");
+  validation.checkId(projectId);
 
-  const contractorCollection = await contractors();
-  const contractor = await contractorCollection.findOne({
-    _id: new ObjectId(contractorId),
-  });
+  const contractor = await getContractor(contractorId);
 
-  const queue = contractor.queue;
-  if (queue.length == 0) {
-    console.log("Empty queue!");
-    throw "Empty queue!";
+  const todo = contractor.todo;
+  let target;
+  for (let project of todo) {
+    if (project.projectId == projectId) {
+      target = project;
+    }
+  }
+  if (!target) {
+    throw `Error: No project found with the given id ${projectId}.`;
+  }
+  if (target.tasks.length == 0) {
+    throw "No more tasks in queue.";
   }
 
-  const task = contractor.queue[0];
-  await contractorCollection.updateOne(
-    { _id: new ObjectId(contractorId) },
-    { $pop: { queue: -1 } }
-  );
-  await contractorCollection.updateOne(
+  const task = target.tasks[0];
+  target.tasks = target.tasks.slice(1);
+
+  for (let i = 0; i < todo.length; i++) {
+    if (todo[i].projectId == projectId) {
+      todo[i] = target;
+    }
+  }
+
+  const contractorCollection = await contractors();
+  const updated = await contractorCollection.updateOne(
     { _id: new ObjectId(contractorId) },
-    { $set: { inProgress: task } }
+    { $set: {todo: todo, inProgress: task } }
   );
+  if (!updated.acknowledged) {
+    throw "Mongo Error: Could not set next task.";
+  }
 
-  const newContractor = await contractorCollection.findOne({
-    _id: new ObjectId(contractorId),
-  });
+  const newContractor = await getContractor(contractorId);
 
-  console.log("In Progress: ", newContractor.inProgress);
-  console.log("Queue: ", newContractor.queue);
-  return true;
+  return await getContractor(contractorId);
 };
 
-const addToInQueue = async (contractorId, task) => {
+const addTaskToQueue = async (contractorId, projectId, task) => {
+  validation.checkNumOfArgs(arguments, 3);
+  validation.checkIsProper(contractorId, "string", "contractorId");
   validation.checkId(contractorId);
+  validation.checkIsProper(projectId, "string", "projectId");
+  validation.checkId(projectId);
   validation.checkForValue(task);
 
+  const contractor = await getContractor(contractorId);
+
+  const todo = contractor.todo;
+  let target;
+  for (let project of todo) {
+    if (project.projectId == projectId) {
+      target = project;
+    }
+  }
+  if (!target) {
+    throw `Error: No project found with the given id ${projectId}.`;
+  }
+  if (target.tasks.length == 0) {
+    throw "Error: No more tasks in queue.";
+  }
```

Comment:
Unnecessary check, shouldn't check if queue is empty if you are trying to add to it